dataflow/gpu-workers/landsat_view.py

```diff
@@ -0,0 +1,388 @@
+    unbounded positive numbers depending on the sensor's exposure.
+    The values are reshaped into (width, height, band), the values are clamped
+    to integers between 0 and 255, and a gamma correction value is applied.
+
+    Args:
+        scene: Landsat 8 scene ID.
+        values: Band values in the shape (band, width, height).
+        min_value: Minimum band value.
+        max_value: Maximum band value.
+        gamma: Gamma correction value.
+
+    Returns:
+        A (scene, pixels) pair. The pixels are Image-ready values.
+    """
+    values = np.array(values, np.float32)
+    logging.info(
+        "{}: preprocess_pixels({}, min={}, max={}, gamma={})".format(
+            scene, values.shape, min_value, max_value, gamma
+        )
+    )
+
+    # Reshape (band, width, height) into (width, height, band).
+    pixels = tf.transpose(values, (1, 2, 0))
+
+    # Rescale to values from 0.0 to 1.0 and clamp them into that range.
+    pixels -= min_value
+    pixels /= max_value
+    pixels = tf.clip_by_value(pixels, 0.0, 1.0)
+
+    # Apply gamma correction.
+    pixels **= 1.0 / gamma
+
+    # Return the pixel values as int8 in the range from 0 to 255,
+    # which is what PIL.Image expects.
+    return scene, tf.cast(pixels * 255.0, dtype=tf.uint8)
+
+
+def save_to_gcs(
+    scene: str, image: Image.Image, output_path_prefix: str, format: str = "JPEG"
+) -> None:
+    """Saves a PIL.Image as a JPEG file in the desired path.
+
+    Args:
+        scene: Landsat 8 scene ID.
+        image: A PIL.Image object.
+        output_path_prefix: Path prefix to save the output files.
+        format: Image format to save files.
+    """
+    filename = os.path.join(output_path_prefix, scene + "." + format.lower())
+    with tf.io.gfile.GFile(filename, "w") as f:
+        image.save(f, format)
+
+
+def run(
+    scenes: List[str],
+    output_path_prefix: str,
+    vis_params: Dict[str, Any],
+    gpu_required: bool,
+    beam_args: Optional[List[str]] = None,
+) -> None:
+    """Load multiple Landsat scenes and render them as JPEG files.
+
+    Args:
+        scenes: List of Landsat 8 scene IDs.
+        output_path_prefix: Path prefix to save the output files.
+        vis_params: Visualization parameters including {rgb_bands, min, max, gamma}.
+        beam_args: Optional list of arguments for Beam pipeline options.
+    """
+    rgb_band_names = vis_params["rgb_band_names"]
+    min_value = vis_params["min"]
+    max_value = vis_params["max"]
+    gamma = vis_params["gamma"]
+
+    options = PipelineOptions(beam_args, save_main_session=True)
+    with beam.Pipeline(options=options) as pipeline:
+        (
+            pipeline
+            | "Create scene IDs" >> beam.Create(scenes)
+            | "Get RGB band paths" >> beam.FlatMap(get_valid_band_paths, rgb_band_names)
+            | "Load RGB band values" >> beam.MapTuple(load_band_values)
+            | "Group RGB band values per scene" >> beam.GroupByKey()
```

Comment:
I there is an inefficiency in parallelism here since we read all the input before performing the grouping by scene and doing downstream processing, while we should be able to process each scene in parallel.

Reply:
I guess we could preprocess each band first, then GroupByKey, and then transpose them into the correct shape. In practical terms, the preprocessing in GPU takes less than a second so it shouldn't make anything slower. And we still need to GBK at some point to make the image.

Reply:
Another thing I noticed is that when we read the raw data, it comes as uint16. After preprocessing we convert it to float32, so it would need to write double the space in temporary writes in Cloud Storage for the snapshots. I noticed that made things quite slower in practice as well.

Reply:
After an offline discussion, it turned out that getting rid of the GroupByKey made the job not having to store the temporary data into Cloud Storage, saving lots of reads/writes. In the end, it turned out to be faster and simpler to have each scene load the three band files sequentially and always keeping them together.